server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/BaseStarTreeBuilder.java

```diff
@@ -317,27 +455,35 @@ private void serializeStarTree(int numSegmentStarTreeDocuments, int numStarTreeD
         );
     }
 
+    /**
+     * Creates the star-tree docValues indices in disk
+     */
     private void createSortedDocValuesIndices(DocValuesConsumer docValuesConsumer, AtomicInteger fieldNumberAcrossStarTrees)
         throws IOException {
-        List<SortedNumericDocValuesWriterWrapper> dimensionWriters = new ArrayList<>();
-        List<SortedNumericDocValuesWriterWrapper> metricWriters = new ArrayList<>();
+        List<DocValuesWriterWrapper<?>> dimensionWriters = new ArrayList<>();
+        List<DocValuesWriterWrapper<?>> metricWriters = new ArrayList<>();
         FieldInfo[] dimensionFieldInfoList = new FieldInfo[numDimensions];
         FieldInfo[] metricFieldInfoList = new FieldInfo[metricAggregatorInfos.size()];
         int dimIndex = 0;
         for (Dimension dim : dimensionsSplitOrder) {
             for (String name : dim.getSubDimensionNames()) {
                 final FieldInfo fi = getFieldInfo(
                     fullyQualifiedFieldNameForStarTreeDimensionsDocValues(starTreeField.getName(), name),
-                    DocValuesType.SORTED_NUMERIC,
+                    dim.getDocValuesType(),
                     fieldNumberAcrossStarTrees.getAndIncrement()
                 );
                 dimensionFieldInfoList[dimIndex] = fi;
-                dimensionWriters.add(new SortedNumericDocValuesWriterWrapper(fi, Counter.newCounter()));
+                if (dim.getDocValuesType().equals(DocValuesType.SORTED_SET)) {
+                    ByteBlockPool.DirectTrackingAllocator byteBlockAllocator = new ByteBlockPool.DirectTrackingAllocator(bytesUsed);
+                    ByteBlockPool docValuesBytePool = new ByteBlockPool(byteBlockAllocator);
+                    dimensionWriters.add(new SortedSetDocValuesWriterWrapper(fi, bytesUsed, docValuesBytePool));
+                } else {
+                    dimensionWriters.add(new SortedNumericDocValuesWriterWrapper(fi, bytesUsed));
```

Comment:
we don't need new Counters for each field?

Reply:
Idea behind this counter is to track memory usage and should be actually used class wide